From 47acdece18b866adb6ede6ef14db37adf06016e1 Mon Sep 17 00:00:00 2001
From: Sergey Fedorov <barracuda@macos-powerpc.org>
Date: Mon, 27 Jul 2026 20:34:05 +0000
Subject: [PATCH 10/16] Implement wp_content_type_manager_v1

Vendor the staging content-type-v1 protocol and accept-and-ignore
set_content_type, matching the get_toplevel_decoration destroy-listener
pattern to reject a second content type object per surface.
---
 .../Compositor/OwlWpContentTypeManagerV1.h    |  37 +++
 .../Compositor/OwlWpContentTypeManagerV1.m    | 214 ++++++++++++++++++
 Sources/Protocol/content-type-v1.xml          | 128 +++++++++++
 Sources/Server/OwlServer.m                    |   2 +
 4 files changed, 381 insertions(+)
 create mode 100644 Sources/Compositor/OwlWpContentTypeManagerV1.h
 create mode 100644 Sources/Compositor/OwlWpContentTypeManagerV1.m
 create mode 100644 Sources/Protocol/content-type-v1.xml

diff --git a/Sources/Compositor/OwlWpContentTypeManagerV1.h b/Sources/Compositor/OwlWpContentTypeManagerV1.h
new file mode 100644
index 0000000..95b73be
--- /dev/null
+++ b/Sources/Compositor/OwlWpContentTypeManagerV1.h
@@ -0,0 +1,37 @@
+/* This file is part of Owl.
+ *
+ * Copyright © 2019-2021 Sergey Bugaev <bugaevc@gmail.com>
+ *
+ * Owl is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published
+ * by the Free Software Foundation, either version 3 of the License,
+ * or (at your option) any later version.
+ *
+ * Owl is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Owl.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#import "OwlGlobal.h"
+#import <Cocoa/Cocoa.h>
+#import <wayland-server.h>
+
+
+/* Implements wp_content_type_manager_v1.
+ *
+ * Owl has no content-type-specific rendering path, so this simply
+ * accepts and ignores the client's hint.
+ */
+@interface OwlWpContentTypeManagerV1 : NSObject <OwlGlobal> {
+    struct wl_resource *_resource;
+}
+
+- (id) initWithResource: (struct wl_resource *) resource;
+
++ (void) addGlobalToDisplay: (struct wl_display *) display;
+
+@end
diff --git a/Sources/Compositor/OwlWpContentTypeManagerV1.m b/Sources/Compositor/OwlWpContentTypeManagerV1.m
new file mode 100644
index 0000000..971d9ce
--- /dev/null
+++ b/Sources/Compositor/OwlWpContentTypeManagerV1.m
@@ -0,0 +1,214 @@
+/* This file is part of Owl.
+ *
+ * Copyright © 2019-2021 Sergey Bugaev <bugaevc@gmail.com>
+ *
+ * Owl is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published
+ * by the Free Software Foundation, either version 3 of the License,
+ * or (at your option) any later version.
+ *
+ * Owl is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Owl.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#import "OwlWpContentTypeManagerV1.h"
+#import "content-type-v1.h"
+#import <wayland-server.h>
+
+
+/* Tracks which wl_surface resources already have a wp_content_type_v1
+ * attached, so a second get_surface_content_type can be rejected. */
+@interface OwlWpContentType : NSObject {
+@public
+    struct wl_resource *_resource;
+    struct wl_resource *_surfaceResource;
+    struct wl_listener _surfaceDestroyListener;
+}
+
+- (id) initWithResource: (struct wl_resource *) resource
+         surfaceResource: (struct wl_resource *) surfaceResource;
+
+@end
+
+@implementation OwlWpContentType
+
+static NSMutableArray *contentTypes;
+
++ (void) initialize {
+    if (contentTypes == nil) {
+        contentTypes = [[NSMutableArray alloc] initWithCapacity: 1];
+    }
+}
+
++ (BOOL) surfaceAlreadyHasContentType: (struct wl_resource *) surfaceResource {
+    for (OwlWpContentType *contentType in contentTypes) {
+        if (contentType->_surfaceResource == surfaceResource) {
+            return YES;
+        }
+    }
+    return NO;
+}
+
+static void surface_destroy_notify(struct wl_listener *listener, void *data) {
+    OwlWpContentType *self = nil;
+    for (OwlWpContentType *contentType in contentTypes) {
+        if (&contentType->_surfaceDestroyListener == listener) {
+            self = contentType;
+            break;
+        }
+    }
+    if (self == nil) {
+        return;
+    }
+
+    self->_surfaceResource = NULL;
+    wl_list_remove(&self->_surfaceDestroyListener.link);
+    wl_list_init(&self->_surfaceDestroyListener.link);
+}
+
+static void content_type_destroy(struct wl_resource *resource) {
+    OwlWpContentType *self = wl_resource_get_user_data(resource);
+    [contentTypes removeObjectIdenticalTo: self];
+    [self release];
+}
+
+static void content_type_destroy_handler(
+    struct wl_client *client,
+    struct wl_resource *resource
+) {
+    wl_resource_destroy(resource);
+}
+
+static void content_type_set_content_type_handler(
+    struct wl_client *client,
+    struct wl_resource *resource,
+    uint32_t content_type
+) {
+    // We have no content-type-specific rendering path; ignore the hint.
+}
+
+static const struct wp_content_type_v1_interface content_type_impl = {
+    .destroy = content_type_destroy_handler,
+    .set_content_type = content_type_set_content_type_handler
+};
+
+- (id) initWithResource: (struct wl_resource *) resource
+         surfaceResource: (struct wl_resource *) surfaceResource
+{
+    _resource = resource;
+    _surfaceResource = surfaceResource;
+    _surfaceDestroyListener.notify = surface_destroy_notify;
+    wl_resource_add_destroy_listener(
+        surfaceResource,
+        &_surfaceDestroyListener
+    );
+    [contentTypes addObject: self];
+
+    wl_resource_set_implementation(
+        resource,
+        &content_type_impl,
+        [self retain],
+        content_type_destroy
+    );
+
+    return self;
+}
+
+- (void) dealloc {
+    if (_surfaceResource != NULL) {
+        wl_list_remove(&_surfaceDestroyListener.link);
+    }
+    [super dealloc];
+}
+
+@end
+
+
+@implementation OwlWpContentTypeManagerV1
+
+static void content_type_manager_destroy(struct wl_resource *resource) {
+    OwlWpContentTypeManagerV1 *self = wl_resource_get_user_data(resource);
+    [self release];
+}
+
+static void content_type_manager_destroy_handler(
+    struct wl_client *client,
+    struct wl_resource *resource
+) {
+    wl_resource_destroy(resource);
+}
+
+static void content_type_manager_get_surface_content_type_handler(
+    struct wl_client *client,
+    struct wl_resource *resource,
+    uint32_t id,
+    struct wl_resource *surface_resource
+) {
+    if ([OwlWpContentType surfaceAlreadyHasContentType: surface_resource]) {
+        wl_resource_post_error(
+            resource,
+            WP_CONTENT_TYPE_MANAGER_V1_ERROR_ALREADY_CONSTRUCTED,
+            "wl_surface already has a wp_content_type_v1"
+        );
+        return;
+    }
+
+    struct wl_resource *content_type_resource = wl_resource_create(
+        client,
+        &wp_content_type_v1_interface,
+        wl_resource_get_version(resource),
+        id
+    );
+    OwlWpContentType *contentType = [OwlWpContentType alloc];
+    [[contentType initWithResource: content_type_resource
+                    surfaceResource: surface_resource] release];
+}
+
+static const struct wp_content_type_manager_v1_interface
+content_type_manager_impl = {
+    .destroy = content_type_manager_destroy_handler,
+    .get_surface_content_type = content_type_manager_get_surface_content_type_handler
+};
+
+- (id) initWithResource: (struct wl_resource *) resource {
+    _resource = resource;
+    wl_resource_set_implementation(
+        resource,
+        &content_type_manager_impl,
+        [self retain],
+        content_type_manager_destroy
+    );
+    return self;
+}
+
+static void content_type_manager_bind(
+    struct wl_client *client,
+    void *data,
+    uint32_t version,
+    uint32_t id
+) {
+    struct wl_resource *resource = wl_resource_create(
+        client,
+        &wp_content_type_manager_v1_interface,
+        version,
+        id
+    );
+    [[[OwlWpContentTypeManagerV1 alloc] initWithResource: resource] release];
+}
+
++ (void) addGlobalToDisplay: (struct wl_display *) display {
+    wl_global_create(
+        display,
+        &wp_content_type_manager_v1_interface,
+        1,
+        NULL,
+        content_type_manager_bind
+    );
+}
+
+@end
diff --git a/Sources/Protocol/content-type-v1.xml b/Sources/Protocol/content-type-v1.xml
new file mode 100644
index 0000000..65a1acb
--- /dev/null
+++ b/Sources/Protocol/content-type-v1.xml
@@ -0,0 +1,128 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<protocol name="content_type_v1">
+  <copyright>
+    Copyright © 2021 Emmanuel Gil Peyrot
+    Copyright © 2022 Xaver Hugl
+
+    Permission is hereby granted, free of charge, to any person obtaining a
+    copy of this software and associated documentation files (the "Software"),
+    to deal in the Software without restriction, including without limitation
+    the rights to use, copy, modify, merge, publish, distribute, sublicense,
+    and/or sell copies of the Software, and to permit persons to whom the
+    Software is furnished to do so, subject to the following conditions:
+
+    The above copyright notice and this permission notice (including the next
+    paragraph) shall be included in all copies or substantial portions of the
+    Software.
+
+    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+    IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+    FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL
+    THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+    LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+    FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+    DEALINGS IN THE SOFTWARE.
+  </copyright>
+
+  <interface name="wp_content_type_manager_v1" version="1">
+    <description summary="surface content type manager">
+      This interface allows a client to describe the kind of content a surface
+      will display, to allow the compositor to optimize its behavior for it.
+
+      Warning! The protocol described in this file is currently in the testing
+      phase. Backward compatible changes may be added together with the
+      corresponding interface version bump. Backward incompatible changes can
+      only be done by creating a new major version of the extension.
+    </description>
+
+    <request name="destroy" type="destructor">
+      <description summary="destroy the content type manager object">
+        Destroy the content type manager. This doesn't destroy objects created
+        with the manager.
+      </description>
+    </request>
+
+    <enum name="error">
+      <entry name="already_constructed" value="0"
+             summary="wl_surface already has a content type object"/>
+    </enum>
+
+    <request name="get_surface_content_type">
+      <description summary="create a new content type object">
+        Create a new content type object associated with the given surface.
+
+        Creating a wp_content_type_v1 from a wl_surface which already has one
+        attached is a client error: already_constructed.
+      </description>
+      <arg name="id" type="new_id" interface="wp_content_type_v1"/>
+      <arg name="surface" type="object" interface="wl_surface"/>
+    </request>
+  </interface>
+
+  <interface name="wp_content_type_v1" version="1">
+    <description summary="content type object for a surface">
+      The content type object allows the compositor to optimize for the kind
+      of content shown on the surface. A compositor may for example use it to
+      set relevant drm properties like "content type".
+
+      The client may request to switch to another content type at any time.
+      When the associated surface gets destroyed, this object becomes inert and
+      the client should destroy it.
+    </description>
+
+    <request name="destroy" type="destructor">
+      <description summary="destroy the content type object">
+        Switch back to not specifying the content type of this surface. This is
+        equivalent to setting the content type to none, including double
+        buffering semantics. See set_content_type for details.
+      </description>
+    </request>
+
+    <enum name="type">
+      <description summary="possible content types">
+        These values describe the available content types for a surface.
+      </description>
+      <entry name="none" value="0">
+        <description summary="no content type applies">
+          The content type none means that either the application has no data
+          about the content type, or that the content doesn't fit into one of
+          the other categories.
+        </description>
+      </entry>
+      <entry name="photo" value="1">
+        <description summary="photo content type">
+          The content type photo describes content derived from digital still
+          pictures and may be presented with minimal processing.
+        </description>
+      </entry>
+      <entry name="video" value="2">
+        <description summary="video content type">
+          The content type video describes a video or animation and may be
+          presented with more accurate timing to avoid stutter. Where scaling
+          is needed, scaling methods more appropriate for video may be used.
+        </description>
+      </entry>
+      <entry name="game" value="3">
+        <description summary="game content type">
+          The content type game describes a running game. Its content may be
+          presented with reduced latency.
+        </description>
+      </entry>
+    </enum>
+
+    <request name="set_content_type">
+      <description summary="specify the content type">
+        Set the surface content type. This informs the compositor that the
+        client believes it is displaying buffers matching this content type.
+
+        This is purely a hint for the compositor, which can be used to adjust
+        its behavior or hardware settings to fit the presented content best.
+
+        The content type is double-buffered state, see wl_surface.commit for
+        details.
+      </description>
+      <arg name="content_type" type="uint" enum="type"
+           summary="the content type"/>
+    </request>
+  </interface>
+</protocol>
diff --git a/Sources/Server/OwlServer.m b/Sources/Server/OwlServer.m
index 7a7fbff..6cab3b6 100644
--- a/Sources/Server/OwlServer.m
+++ b/Sources/Server/OwlServer.m
@@ -25,6 +25,7 @@
 #import "OwlZxdgShellV6.h"
 #import "OwlXdgWmBase.h"
 #import "OwlZxdgDecorationManagerV1.h"
+#import "OwlWpContentTypeManagerV1.h"
 #import "OwlSeat.h"
 #import "OwlOutput.h"
 #import "OwlWlDataDeviceManager.h"
@@ -127,6 +128,7 @@
     [OwlZxdgShellV6 addGlobalToDisplay: _display];
     [OwlXdgWmBase addGlobalToDisplay: _display];
     [OwlZxdgDecorationManagerV1 addGlobalToDisplay: _display];
+    [OwlWpContentTypeManagerV1 addGlobalToDisplay: _display];
     [OwlWlDataDeviceManager addGlobalToDisplay: _display];
     [OwlSeat addGlobalToDisplay: _display];
     [OwlOutput addGlobalToDisplay: _display];
-- 
2.43.0

